tuf/api/metadata.py

```diff
@@ -1643,25 +1665,78 @@ def from_dict(cls, delegations_dict: Dict[str, Any]) -> "Delegations":
-            if new_role.name in roles_res:
-                raise ValueError(f"Duplicate role {new_role.name}")
-            roles_res[new_role.name] = new_role
+        roles = delegations_dict.pop("roles", None)
+        roles_res: Optional[Dict[str, DelegatedRole]] = None
+
+        if roles is not None:
+            roles_res = {}
+            for role_dict in roles:
+                new_role = DelegatedRole.from_dict(role_dict)
+                if new_role.name in roles_res:
+                    raise ValueError(f"Duplicate role {new_role.name}")
+                roles_res[new_role.name] = new_role
+
+        succinct_roles_dict = delegations_dict.pop("succinct_roles", None)
+        succinct_roles_info = None
+        if succinct_roles_dict is not None:
+            succinct_roles_info = SuccinctRoles.from_dict(succinct_roles_dict)
+
         # All fields left in the delegations_dict are unrecognized.
-        return cls(keys_res, roles_res, delegations_dict)
+        return cls(keys_res, roles_res, succinct_roles_info, delegations_dict)
 
     def to_dict(self) -> Dict[str, Any]:
         """Returns the dict representation of self."""
         keys = {keyid: key.to_dict() for keyid, key in self.keys.items()}
-        roles = [role_obj.to_dict() for role_obj in self.roles.values()]
-        return {
+        res_dict: Dict[str, Any] = {
             "keys": keys,
-            "roles": roles,
             **self.unrecognized_fields,
         }
+        if isinstance(self.roles, dict):
+            roles = [role_obj.to_dict() for role_obj in self.roles.values()]
+            res_dict["roles"] = roles
+        elif isinstance(self.succinct_roles, SuccinctRoles):
+            res_dict["succinct_roles"] = self.succinct_roles.to_dict()
+        return res_dict
+
+    def find_delegation(self, target_filepath: str) -> Optional[str]:
+        """Find the delegated rolename based on the given ``target filepath``.
+
+        If the given ``target_filepath`` is not in any of the paths that
+        ``Delegations`` is trusted to provide it returns None.
+
+        Args:
+            target_filepath: URL path to a target file, relative to a base
+                targets URL.
+        """
+        if isinstance(self.roles, dict):
+            for role in self.roles.values():
+                if role.is_delegated_path(target_filepath):
+                    return role.name
+
+        elif self.succinct_roles is not None:
+            return self.succinct_roles.find_bin(target_filepath)
+
+        return None
+
+    def get_all_delegations(
+        self, target_filepath: str
+    ) -> Iterator[Tuple[str, bool]]:
+        """Given ``target_filepath`` get names and terminating status of all
+        delegations who are responsible for it.
+
+        Args:
+            target_filepath: URL path to a target file, relative to a base
+                targets URL.
+        """
+        if isinstance(self.roles, dict):
```

Comment:
That's a good catch.
I thought that I need to verify that `self.roles` is actually a `Dict` in order for `mypy` to understands why I am using dictionary methods, but you are right that `self.roles` is defined clearly as `Optional[Dict]` and it shouldn't be something else rather than `None` or `Dict`.